runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java

```diff
@@ -312,18 +316,22 @@ public String toString() {
   }
 
   /**
-   * Cancels the workflow on the first error message it sees.
+   * Monitors job log output messages for errors.
    *
    * <p>Creates an error message representing the concatenation of all error messages seen.
    */
-  private static class CancelWorkflowOnError implements JobMessagesHandler {
+  private static class ErrorMonitorMessagesHandler implements JobMessagesHandler {
     private final DataflowPipelineJob job;
     private final JobMessagesHandler messageHandler;
     private final StringBuffer errorMessage;
-    private CancelWorkflowOnError(DataflowPipelineJob job, JobMessagesHandler messageHandler) {
+    private volatile boolean hasSeenError;
+
```

Comment:
TBH I'd be happier if this gather the messages and separated formatting from that, rather than tacking together the message text as it goes.